File: mail-assistant/src/summary.ts
/**
 * 日次集計（純関数 + 通知）。
 *
 * 18:05 頃にその日の処理結果を集計する。
 * メール送信は実装しない（自動送信を作らない方針）。出力先はログか Google Chat のみ。
 */
import type { Config } from './config.js';
import type { ProcessingRecord } from './types.js';

export interface DailyStats {
  readonly date: string;
  readonly examined: number;
  readonly drafted: number;
  readonly review: number;
  readonly noReply: number;
  readonly errors: number;
  readonly important: number;
  readonly injectionSuspected: number;
  readonly dryRunOnly: boolean;
}

/** 履歴レコードから当日の集計を作る。 */
export function aggregate(date: string, records: readonly ProcessingRecord[]): DailyStats {
  let drafted = 0;
  let review = 0;
  let noReply = 0;
  let errors = 0;
  let important = 0;
  let injectionSuspected = 0;
  let anyDraftId = false;

  for (const r of records) {
    switch (r.classification) {
      case 'REPLY_REQUIRED':
        drafted += 1;
        break;
      case 'REVIEW_REQUIRED':
        review += 1;
        break;
      case 'NO_REPLY_REQUIRED':
        noReply += 1;
        break;
    }
    if (r.error !== '') errors += 1;
    if (r.important) important += 1;
    if (r.injectionSuspected) injectionSuspected += 1;
    if (r.draftId !== '') anyDraftId = true;
  }

  return {
    date,
    examined: records.length,
    drafted,
    review,
    noReply,
    errors,
    important,
    injectionSuspected,
    // 下書き ID が1つも無ければドライラン（または該当なし）だったと見なす
    dryRunOnly: !anyDraftId,
  };
}

/** 通知用テキスト。個人情報は含めない（件数のみ）。 */
export function formatSummary(stats: DailyStats, config: Config): string {
  const lines = [
    `🗂 メール返信下書きアシスタント 日次集計 ${stats.date}`,
    `対象アカウント: ${config.targetEmail}`,
    '',
    `確認したメール数: ${stats.examined}`,
    `返信下書きを作成: ${stats.drafted}`,
    `要確認: ${stats.review}`,
    `返信不要: ${stats.noReply}`,
    `エラー: ${stats.errors}`,
  ];
  if (stats.important > 0) lines.push(`重要メール（請求・契約・セキュリティ等）: ${stats.important}`);
  if (stats.injectionSuspected > 0) {
    lines.push(`プロンプトインジェクションの疑い: ${stats.injectionSuspected}`);
  }
  if (config.dryRun) {
    lines.push('', '※ ドライラン中です。Gmail への下書き作成・ラベル付与は行っていません。');
  } else if (stats.dryRunOnly && stats.drafted > 0) {
    lines.push('', '※ 下書き作成が0件でした。ラベル・権限設定を確認してください。');
  }
  return lines.join('\n');
}

File: mail-assistant/summary.py
"""日次集計。

メール送信は行わない（そもそもコネクタに送信ツールが無い）。
出力は実行ログのみで、Routine の完了通知（プッシュ／メール）に乗って手元へ届く。
"""
from __future__ import annotations


def aggregate(iso_date: str, records: list[dict]) -> dict:
    """当日の処理履歴から件数を集計する。"""
    stats = {
        "date": iso_date,
        "examined": len(records),
        "drafted": 0,
        "review": 0,
        "noReply": 0,
        "errors": 0,
        "important": 0,
        "injectionSuspected": 0,
        "draftsCreated": 0,
    }
    for record in records:
        classification = record.get("classification")
        if classification == "REPLY_REQUIRED":
            stats["drafted"] += 1
        elif classification == "REVIEW_REQUIRED":
            stats["review"] += 1
        elif classification == "NO_REPLY_REQUIRED":
            stats["noReply"] += 1
        if record.get("error"):
            stats["errors"] += 1
        if record.get("important"):
            stats["important"] += 1
        if record.get("injectionSuspected"):
            stats["injectionSuspected"] += 1
        if record.get("draftId"):
            stats["draftsCreated"] += 1
    return stats


def format_summary(stats: dict, config: dict) -> str:
    """通知用テキスト。件数のみで個人情報を含めない。"""
    lines = [
        f"📧 メール返信下書きアシスタント 日次集計 {stats['date']}",
        f"対象アカウント: {config['targetEmail']}",
        "",
        f"確認したメール数: {stats['examined']}",
        f"返信下書きを作成: {stats['drafted']}（実際に作成した下書き: {stats['draftsCreated']}）",
        f"要確認: {stats['review']}",
        f"返信不要: {stats['noReply']}",
        f"エラー: {stats['errors']}",
    ]
    if stats["important"]:
        lines.append(f"重要メール（請求・契約・セキュリティ等）: {stats['important']}")
    if stats["injectionSuspected"]:
        lines.append(f"プロンプトインジェクションの疑い: {stats['injectionSuspected']}")

    if config.get("dryRun", True):
        lines += ["", "※ ドライラン中です。下書き作成・ラベル付与は行っていません。"]
    elif stats["drafted"] and not stats["draftsCreated"]:
        lines += ["", "※ 返信必要と判定したのに下書きが0件です。ラベル・権限設定を確認してください。"]
    if stats["review"]:
        lines += ["", f"「{config.get('labels', {}).get('review', 'AI要確認')}」ラベルを確認してください。"]
    return "\n".join(lines)
